File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        
        if nums[l] == target:
            return l
        elif nums[r] == target:
            return r
        
        def binary_search(nums: List[int], target: int, left: int, right: int) -> int:
            while left < right:
                mid = left + (right - left) // 2
                
                if nums[mid] < target:
                    left = mid + 1
                elif nums[mid] > target:
                    right = mid
                else:
                    return mid
                
            return -1
        
        # array was not rotated -> find from first to last
        if nums[l] <= nums[r]:
            return binary_search(nums, target, l, r)
        
        while l < r:
            m = l + (r - l) // 2
            if nums[m] < nums[r]:
                r = m
            else:
                l = m + 1
            
        # target < nums[0] -> it is in the right of pivot -> find from pivot to end
        if target < nums[0]:
            return binary_search(nums, target, l, len(nums) - 1)
        
        # else -> it is in the left of pivot -> find from begin to pivot
        return binary_search(nums, target, 0, l)
        
            

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    auto binary_search(std::vector<int> &nums, int target, int l, int r) -> int {
        while (l < r) {
            int m = l + (r - l) / 2;
            
            if (nums[m] > target)
                r = m;
            else if (nums[m] < target)
                l = m + 1;
            else
                return m;
        }
        return -1;
    }
    
    int search(vector<int>& nums, int target) {
        if (!nums.size())
            return -1;
        
        int l = 0,
            r = nums.size() - 1;
        
        if (nums[l] == target || nums[r] == target) {
            return (nums[l] == target)? l:r;
        }
        
        //not rotated
        if (nums[l] <= nums[r]) {
            return binary_search(nums, target, l, r);
        }

        // rotated -> *nums.begin() > *nums.end()
        while (l < r) {
            int m = l + (r - l)/2,
                mid = nums[m];
            
            if (mid < nums.back())
                r = m;
            else
                l = m + 1;
        }
        
        return nums[0] > target ? binary_search(nums, target, l, nums.size() - 1):binary_search(nums, target, 0, l);
    }
};